src/cli/cli.cpp

```diff
@@ -4615,8 +4615,7 @@ otError Interpreter::ProcessDiag(uint8_t aArgsLength, Arg aArgs[])
     char    output[OPENTHREAD_CONFIG_DIAG_OUTPUT_BUFFER_SIZE];
 
     // all diagnostics related features are processed within diagnostics module
-    output[0]                  = '\0';
-    output[sizeof(output) - 1] = '\0';
+    memset(output, 0, sizeof(output));
```

Comment:
Shouldn't the client ensure that what they write to the output buffer is null-terminated? In what case do we expect the client to not null terminte?

That said, I am not opposed to initializing the output buffer to all zeros.

Reply:
Besides, we actually don't define how `output` should be initialized in API `otDiagProcessCmd`. Technically speaking, we should not initialize `output` at all. 
It's `Diags::ProcessCmd`'s responsiblity to make sure the output is null-terminated?

Reply:
It looks like currently the implementation of `otDiagProcessCmd` doesn't ensure the output will be null-terminated. For example, in most cases `AppendErrorResult` is called at the end and it doesn't append a `\0` to the output. 

I think the problem is: do we want to let `otDiagProcessCmd` to ensure the output is null-terminated (then we also need to require `otPlatDiagProcess` to make sure that)? And we should add document and update the implementation for those APIs. Or we require the caller of the API to pass an all-zero buffer to `otDiagProcessCmd`. Both approaches make sense to me. Thoughts? @jwhui @simonlingoogle

Reply:
I think `otDiagProcessCmd` should make sure the output is null-terminated as long as it returns `kErrorNone`.

Reply:
> I think `otDiagProcessCmd` should make sure the output is null-terminated as long as it returns kErrorNone.

I fully agree. +1 on this.